src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -6518,289 +6524,265 @@ int SketchObject::deleteUnusedInternalGeometry(int GeoId, bool delgeoid)
+        if (cpGeoId == GeoEnum::GeoUndef) {
+            continue;
         }
 
-        for (it = knotgeoids.begin(), ita = kassociatedconstraints.begin();
-             it != knotgeoids.end() && ita != kassociatedconstraints.end();
-             ++it, ++ita) {
-            if ((*it) != -1) {
-                // look for a point at geoid index
-                for (std::vector<Sketcher::Constraint*>::const_iterator itc = vals.begin();
-                     itc != vals.end();
-                     ++itc) {
-                    if ((*itc)->Second == (*it) || (*itc)->First == (*it)
-                        || (*itc)->Third == (*it)) {
-                        (*ita)++;
-                    }
-                }
-
-                if ((*ita) < 2) {// IA
-                    delgeometries.push_back((*it));
-                }
+        // look for a circle at geoid index
+        for (auto const& constr : vals) {
+            if (constr->Type == Sketcher::InternalAlignment
+                || constr->Type == Sketcher::Weight
+                || !constr->involvesGeoId(cpGeoId)) {
+                continue;
+            }
+            if (constr->Type != Sketcher::Equal) {
+                ++numConstr;
+                continue;
+            }
+            bool firstIsInCPGeoIds = std::any_of(poleGeoIdsAndConstraints.begin(),
+                                                 poleGeoIdsAndConstraints.end(),
+                                                 [&constr](const auto& _pair) {
+                                                     return _pair.first == constr->First;
+                                                 });
+            bool secondIsInCPGeoIds = std::any_of(poleGeoIdsAndConstraints.begin(),
+                                                  poleGeoIdsAndConstraints.end(),
+                                                  [&constr](const auto& _pair) {
+                                                      return _pair.first == constr->Second;
+                                                  });
+            // the equality constraint constrains a pole but it is not interpole
+            if (firstIsInCPGeoIds != secondIsInCPGeoIds) {
+                ++numConstr;
             }
+            // We do not ignore weight constraints as we did with radius constraints,
+            // because the radius magnitude no longer makes sense without the B-Spline.
         }
 
+        if (numConstr < 1) { // IA
+            delgeometries.push_back(cpGeoId);
+        }
+    }
 
-        if (delgeoid)
-            delgeometries.push_back(GeoId);
+    for (auto& [kGeoId, numConstr] : knotGeoIdsAndConstraints) {
+        if (kGeoId == GeoEnum::GeoUndef) {
+            continue;
+        }
 
-        int ndeleted = delGeometriesExclusiveList(delgeometries);
+        // look for a point at geoid index
+        numConstr = std::count_if(vals.begin(), vals.end(), [&kGeoId](const auto& constr) {

```

Comment:
@chennes is there something wrong with using "block variable" in an internal lambda?

Reply:
No, it's just confusing the linter, which then doesn't see the variable in the outer scope and produces an incorrect diagnostic in this case. There's no problem with this code.

Reply:
Perbaps not uodated for c++17?